From 015e8f570cfe25548fc0491406d0c99a515aa407 Mon Sep 17 00:00:00 2001
From: Aihua Xu <aihuaxu@apache.org>
Date: Fri, 26 Feb 2016 14:34:46 -0500
Subject: [PATCH 0523/1164] CDH-37847: HIVE-13160: HS2 unable to load UDFs on
 startup when HMS is not ready (Reviewed by
 Yongzhi Chen, Sergey Shelukhin)

Change-Id: I34f1c6725698d8749dd095d2eca4e7f040e5607a
---
 .../org/apache/hadoop/hive/ql/metadata/Hive.java   |    7 ++++---
 1 file changed, 4 insertions(+), 3 deletions(-)

diff --git a/ql/src/java/org/apache/hadoop/hive/ql/metadata/Hive.java b/ql/src/java/org/apache/hadoop/hive/ql/metadata/Hive.java
index d94f494..665c1f0 100644
--- a/ql/src/java/org/apache/hadoop/hive/ql/metadata/Hive.java
+++ b/ql/src/java/org/apache/hadoop/hive/ql/metadata/Hive.java
@@ -166,7 +166,7 @@ public synchronized void remove() {
   private final static int REG_FUNCS_NO = 0, REG_FUNCS_DONE = 2, REG_FUNCS_PENDING = 1;
 
   // register all permanent functions. need improvement
-  private void registerAllFunctionsOnce() {
+  private void registerAllFunctionsOnce() throws HiveException {
     boolean breakLoop = false;
     while (!breakLoop) {
       int val = didRegisterAllFuncs.get();
@@ -195,11 +195,12 @@ private void registerAllFunctionsOnce() {
     }
     try {
       reloadFunctions();
+      didRegisterAllFuncs.compareAndSet(REG_FUNCS_PENDING, REG_FUNCS_DONE);
     } catch (Exception e) {
       LOG.warn("Failed to register all functions.", e);
+      didRegisterAllFuncs.compareAndSet(REG_FUNCS_PENDING, REG_FUNCS_NO);
+      throw new HiveException(e);
     } finally {
-      boolean result = didRegisterAllFuncs.compareAndSet(REG_FUNCS_PENDING, REG_FUNCS_DONE);
-      assert result;
       synchronized (didRegisterAllFuncs) {
         didRegisterAllFuncs.notifyAll();
       }
-- 
1.7.9.5

